### PR TITLE
nn.Transformer

### PR DESCRIPTION
**Pull request by zhangguanheng66**
Create a PR for comments. The model is still WIP but I want to have some feedbacks before moving too far. The transformer model depends on several modules, like MultiheadAttention (landed).

Transformer is implemented based on the paper (https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf). Users have the flexibility to build a transformer with self-defined and/or built-in components (i.e encoder, decoder, encoder_layer, decoder_layer). Users could use Transformer class to build a standard transformer model and modify sub-layers as needed.

Add a few unit tests for the transformer module, as follow:
TestNN.test_Transformer_cell
TestNN.test_transformerencoderlayer
TestNN.test_transformerdecoderlayer
TestNN.test_transformer_args_check
TestScript.test_scriptmodule_transformer_cuda

There is another demonstration example for applying transformer module on the word language problem. https://github.com/pytorch/examples/pull/555

### CODE REVIEW

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+from .. import functional as F
+from .module import Module
+from .activation import MultiheadAttention
+from .container import ModuleList
+import numpy as np
+from ..init import xavier_uniform_
+from .dropout import Dropout
+from .linear import Linear
+from .normalization import LayerNorm
+
```
Comment (SsnL) [inline torch/nn/modules/transformer.py]: this likely breaks lint

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+class TransformerBase(Module):
```
Comment (SsnL) [inline torch/nn/modules/transformer.py]: do you really need a base when there is only one subclass?

File: torch/nn/modules/transformer.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        memory = self.encoder(src, src_mask)
+        output = self.decoder(tgt, memory, tgt_mask, memory_mask)
+        return output
+
+    def generate_square_subsequent_mask(self, sz):
+        r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
+            Unmasked positions are filled with float(0.0).
+        """
+        attn_shape = (sz, sz)
+        mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
```
Comment (SsnL) [inline torch/nn/modules/transformer.py]: we don't have hard dependency on numpy

Comment (SsnL) [reply torch/nn/modules/transformer.py]: this function also seems inefficient and unused everywhere

File: torch/nn/modules/__init__.py [hunk 4]
```diff
@@ -27,6 +27,8 @@
 from .distance import PairwiseDistance, CosineSimilarity
 from .fold import Fold, Unfold
 from .adaptive import AdaptiveLogSoftmaxWithLoss
+from .transformer import TransformerBase, TransformerEncoder, TransformerDecoder, \
```
Comment (SsnL) [inline torch/nn/modules/__init__.py]: Why are the modules other than `Transformer` exposed?

**Comment by SsnL** [comment]
Nice progress. But I have several comments and I also don't see the test and doc entries.

**Comment by zhangguanheng66** [comment]
> Nice progress. But I have several comments and I also don't see the test and doc entries.

We haven't decided a unit test for the transformer model. I will add the doc entries.

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: Will fix it.

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: This base class just lays out the encoder-decoder architecture.

File: torch/nn/modules/__init__.py (hunk 4 above)
Comment (zhangguanheng66) [reply torch/nn/modules/__init__.py]: sometime users may use parts of the transformer model, so the encoder/decoder  modules are exposed. I agree TransformerBase should not be exposed.

File: torch/nn/modules/transformer.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
+import torch
+import copy
+from .. import functional as F
+from .module import Module
+from .activation import MultiheadAttention
+from .container import ModuleList
+import numpy as np
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Remove this once you've deleted the numpy usage. We don't have a hard dependency on numpy.

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: Let's remove this until we need an abstraction. Just like @SsnL mentioned it's only used in a single class. It also doesn't do more than add those two fields.

File: torch/nn/modules/transformer.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    Examples::
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab)
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab, nhead=16, num_encoder_layers=12)
+    """
+
+    def __init__(self, d_model=512, nhead=8, num_encoder_layers=6,
+                 num_decoder_layers=6, dim_feedforward=2048, dropout=0.1):
+        encoder_layer = TransformerEncoderLayer(d_model, nhead, dim_feedforward, dropout)
+        encoder_norm = LayerNorm(d_model) 
+        encoder = TransformerEncoder(encoder_layer, num_encoder_layers, encoder_norm)
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: We should make encoder and decoder keyword arguments and create this if they're set to None. This allows users to pass in custom code.

File: torch/nn/modules/transformer.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        self.linear2 = Linear(dim_feedforward, d_model)
+
+        self.norm1 = LayerNorm(d_model)
+        self.norm2 = LayerNorm(d_model)
+        self.norm3 = LayerNorm(d_model)
+        self.dropout1 = Dropout(dropout)
+        self.dropout2 = Dropout(dropout)
+        self.dropout3 = Dropout(dropout)
+
+    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: For all the forward functions, please move the compute into [functional.py](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1391). We separate the computations from the state. You can call the function you're adding something like "transformer_decoder_layer". Also, try annotating it with "@weak_script" just like the example I'm referencing above.

**Comment by cpuhrsch** [comment]
Looking much better! Just as @SsnL mentioned, please add unit tests.

File: torch/nn/modules/transformer.py (hunk 7 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: Mitigate them soon.

File: test/test_nn.py [hunk 8]
```diff
@@ -5228,6 +5228,129 @@ def test_LSTM_cell(self):
...
+        # set constant weights of the model
+        for idx, p in enumerate(model.parameters()):
+            x = p.data
+            sz = x.view(-1).size(0)
+            shape = x.shape
+            x = torch.arange(0, sz).float().view(shape)/100
+            p.data.copy_(x)
+
+        # deterministic input
+        encoder_input = torch.Tensor([[[0.7462, 0.6653, 0.5679, 0.4891],
```
Comment (cpuhrsch) [inline test/test_nn.py]: I think you can add more with even smaller and simpler inputs

Comment (cpuhrsch) [reply test/test_nn.py]: smaller here means "smaller tensors" and simpler means "integers"

**Comment by cpuhrsch** [comment]
Pinging @srush @kyunghyuncho @myleott @glample for additional feature requests if any.

**Comment by cpuhrsch** [comment]
Pinging @mansimov @jasonleeinf  @myleott @fmassa for additional feature requests / review.

**Comment by cpuhrsch** [comment]
Pinging @stephenroller @douwekiela for additional feature requests / review.

**Comment by zhangguanheng66** [comment]
We also have a PR for applying torch.nn.Transformer in word language problem (https://github.com/pytorch/examples/pull/555). @myleott

File: torch/nn/modules/transformer.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        r"""Pass the input through the endocder layer.
+
+        Args:
+            src: the sequnce to the encoder layer (required).
+            src_mask: the mask for the src sequence (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        src2 = self.norm1(src)
```
Comment (stephenroller) [inline torch/nn/modules/transformer.py]: You should note that there's some debate about whether these go.

Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: One option is to check before calling the norm function.

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: @stephenroller - do you mean there's debate around what kind of normalization to use or any at all?

@zhangguanheng66 - the docs should mention that we use norm, but I don't think we should make that optional / replaceable for now (unless we get many more comments).

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: It might be easier to reimplement the EncoderLayer than making this one more complicated. But maybe I'm wrong.

Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: We should finally write a post to review the nn.Transformer module and clearly lay out the architecture, including norm.

Comment (stephenroller) [reply torch/nn/modules/transformer.py]: sorry I wrote whether but I meant "where".

Some variations put them before the selfattn/relu. Some put them after.

Comment (douwekiela) [reply torch/nn/modules/transformer.py]: Maybe we should benchmark a bunch of alternatives and pick the best one?

Comment (stephenroller) [reply torch/nn/modules/transformer.py]: Myle and Guillaume have for a bunch of things already. I’ve found it makes a *small* difference in my experiments but usually had too many confounding variables to make a clear recommendation.

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: I suggest we do that in a separate PR since it's not a necessary feature to get the basic featureset.

Comment (stephenroller) [reply torch/nn/modules/transformer.py]: Just to actually ground it in code:

https://github.com/pytorch/fairseq/blob/dffb167449cf0e3ead73b4a67e84f222311b0ff5/fairseq/models/transformer.py#L532-L557

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: Ok, let's standardize on ["Attention is all you need"](https://arxiv.org/abs/1706.03762) for now and mention that in the docs. Later on we can extend the behavior to match the implementation of other Transformer implementations with corresponding paper references.

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: Also see [a comment in fairseq mentioning this](https://github.com/pytorch/fairseq/blob/dffb167449cf0e3ead73b4a67e84f222311b0ff5/fairseq/models/transformer.py#L570).

**Comment by cpuhrsch** [comment]
cc @BIT-silence @kimishpatel

**Comment by cpuhrsch** [comment]
Travis lint failures are relevant and need to be resolved (lines too long).

File: torch/nn/functional.py [hunk 10]
```diff
@@ -3262,3 +3263,147 @@ def _in_proj_v(weight, bias, embed_dim, value):
         attn_output_weights = None
 
     return attn_output, attn_output_weights
+
+@weak_script
+def transformer_encoder_layer_forward(src,                                  # type: Tensor
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: Add to the docs that this is an unstable API, because we'll need to add support for slightly different standard transformers (e.g. BERT) and this is based on "All you need is attention" (which we should also cite and reference).

File: torch/nn/functional.py [hunk 11]
```diff
@@ -3262,3 +3263,147 @@ def _in_proj_v(weight, bias, embed_dim, value):
...
+    src2 = relu(src2)
+    src2 = dropout(src2, dropout_p, dropout_training, dropout_inplace)
+    src2 = linear(src2, linear2_weight, linear2_bias)
+    src2 = dropout(src2, dropout2_p, dropout2_training, dropout2_inplace)
+    src = src + src2
+
+    return src
+
+@weak_script
+def transformer_decoder_layer_forward(tgt,                                  # type: Tensor
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: Add to the docs that this is an unstable API, because we'll need to add support for slightly different standard transformers (e.g. BERT) and this is based on "All you need is attention" (which we should also cite and reference).

File: torch/nn/modules/transformer.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+from .activation import MultiheadAttention
+from .container import ModuleList
+from ..init import xavier_uniform_
+from .dropout import Dropout
+from .linear import Linear
+from .normalization import LayerNorm
+
+class Transformer(Module):
+    r"""A transformer model. User is able to modified the attributes as needed. The architechture
+        is based on the paper "Attention Is All You Need".
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Add full citation

File: torch/nn/modules/transformer.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+            raise RuntimeError("the batch number of src and tgt must be equal")
+
+        if src.size(2) != self.d_model or tgt.size(2) != self.d_model:
+            raise RuntimeError("the feature number of src and tgt must be equal to d_model")
+
+        memory = self.encoder(src, src_mask)
+        output = self.decoder(tgt, memory, tgt_mask, memory_mask)
+        return output
+
+    def generate_square_subsequent_mask(self, sz):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Doesn't seem to be used by models - maybe we want to remove it?

File: torch/nn/modules/transformer.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+
+        for i in range(self.num_layers):
+            output = self.layers[i](output, memory, tgt_mask, memory_mask)
+
+        if self.norm:
+            output = self.norm(output)
+
+        return output
+
+class TransformerEncoderLayer(Module):
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Add in details on standardization choice from discussion below.

File: test/test_jit.py [hunk 15]
```diff
@@ -12572,6 +12572,45 @@ def forward(self, q, k, v):
         # print(torch.allclose(jit_out, py_out, atol=5e-4, rtol=1e-4))
         self.assertTrue(torch.allclose(jit_out, py_out, atol=5e-4, rtol=1e-4))
 
+    @unittest.skipIf(not RUN_CUDA, "no CUDA")
```
Comment (cpuhrsch) [inline test/test_jit.py]: Why no CUDA?

File: torch/nn/modules/transformer.py [hunk 16]
```diff
@@ -235,6 +240,8 @@ def forward(self, src, src_mask=None):
 
 class TransformerDecoderLayer(Module):
     r"""TransformerDecoderLayer is made up of self-attn, multi-head-attn and feedforward network.
+        This standard decoder layer is based on the paper "Attention Is All You Need".
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Reference the note with the full citation for completion.

File: torch/nn/modules/transformer.py (hunk 12 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: added.

File: torch/nn/modules/transformer.py (hunk 16 above)
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: added

**Comment by zhangguanheng66** [comment]
fix https://github.com/pytorch/pytorch/issues/10459

File: torch/nn/modules/transformer.py [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+        num_layers: the number of sub-encoder-layers in the encoder (required).
+        norm: the layer normalization component (optional).
+
+    Examples::
+        >>> transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers)
+    """
+
+    def __init__(self, encoder_layer, num_layers, norm=None):
+        super(TransformerEncoder, self).__init__()
+        self.layers = _get_clones(encoder_layer, num_layers)
```
Comment (DNGros) [inline torch/nn/modules/transformer.py]: Is it intended behavior to have weight sharing? In a standard transformer (like from Attention is All You Need), I don't believe the weights are shared. I understand there has been followup work showing weight sharing can be beneficial in some cases, but if this is intended behavior, it might be useful to specify this difference from the paper's implementation in docs after the Vaswani et al. citation and cite the followup work.

-----

As a broader point: 
This seems to get at the fact that a Transformer is a rather high level component, without really consensus yet on its architecture with a lot decisions likely domain specific (weight sharing?, beam search decoder?, convolutions between layers?, output the intermediate attentions? hierarchy / grouping components of the task?, sparsity?... and many more ideas that are being rapidly published) 

I don't mean to try and start needless debates, or to at all imply this Transformer code is not useful, but I will add in my thoughts that from my limited perspective that maybe Transformers might better belong in the contrib module or in the docs as a example that people can modify to meet their needs. The torch.nn.Modules module seems (at least currently) to be for more foundational components that can be composed into larger models, not full model architectures themselves.  

There has already been fairly extensive discussion on this in #10459 , and it seemed like the consensus there was to focus on things like MultiHeadedAttention or PositionalEncoding for core and keep full architectures separate in the codebase.

Comment (DNGros) [reply torch/nn/modules/transformer.py]: As an addendum to that second part: Thinking about this more I could see how a TransformerEncoder (or like StackedSelfAttention or RecurrentSelfAttention) could maybe be considered a primitive component which could lend itself to maybe eventually being low level optimized and composed into novel things (though not really that much more primitive than something like a ResNet-block). However, when it starts seeming particularly high level, and full-archetecture-y is the inclusion of a prepackaged encoder and decoder Transformer for seq2seq, which, without a lot of additional components, seems likely would not meet all needs and could not be easily adapted/composed. 
I don't know, and I don't really have any reason to hold strong opinions here. Feel free to dismiss this second part without real justification...

Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: @DNGros Thanks for the comments. To your first question, it's not supposed to share the weights between layers. Actually, if you take a quick look at _get_clones function, it calls a "deepcopy" function.

For your second comment, we do realize there are many ongoing discussions about the transformer models and several variants in different domains. We want to provide a baseline for the research community and startups when people don't want to code from scratch. This module could help people try some preliminary ideas for the fast delivery. For some advanced users, we expect them to develop a specific transformer model, and they could possibly use this module as a reference or benchmark case. 

As you suggest, we try to make the model highly "modularized". People can use nn.Tranformer, nn.TransformerEncoder, or EVEN nn.TransformerEncoderLayer, as needed. There is actually a word language example (https://github.com/pytorch/examples/pull/555) where we use nn.TransformerEncoder as the seq2seq model.

In the future, if we see a variant of the transformer model requested widely by the community, we will continuously implement them in our framework. I see a baseline model could benefit more users as we optimize the module performance in the future.

Comment (DNGros) [reply torch/nn/modules/transformer.py]: @zhangguanheng66 Ok, sorry. I missed that.

Thanks for clarifying!

**Review by cpuhrsch — APPROVED** [review]
Merge at will
